lsm.go

```diff
@@ -0,0 +1,352 @@
+package frostdb
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/apache/arrow/go/v14/arrow"
+	"github.com/apache/arrow/go/v14/arrow/util"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/polarsignals/frostdb/dynparquet"
+	"github.com/polarsignals/frostdb/parts"
+	"github.com/polarsignals/frostdb/query/logicalplan"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// LSM is a log-structured merge-tree like index. It is implemented as a single linked list of parts.
+//
+// Arrow records are always added to the L0 list. When a list reaches it's configured max size it is compacted
+// calling the levels Compact function and is then added as a new part to the next level.
+//
+// [L0]->[record]->[record]->[L1]->[record/parquet]->[record/parquet] etc.
+type LSM struct {
+	sync.RWMutex
+	compacting *atomic.Bool
+
+	prefix  string
+	levels  *Node
+	sizes   []*atomic.Int64
+	configs []*LevelConfig
+	schema  *dynparquet.Schema
+
+	logger  log.Logger
+	metrics *LSMMetrics
+}
+
+// LSMMetrics are the metrics for an LSM index.
+type LSMMetrics struct {
+	reg                prometheus.Registerer
+	Compactions        *prometheus.CounterVec
+	LevelSize          *prometheus.GaugeVec
+	CompactionDuration prometheus.Histogram
+}
+
+// LevelConfig is the configuration for a level in the LSM tree.
+// The Level is the sentinel node that represents the level.
+// The MaxSize is the maximum size in bytes that the level can reach before it triggers compaction into the next level.
+// The Compact function is called when the level reaches it's max size. NOTE: that this is not yet implemtened and the database will rotate the full block as normal.
+type LevelConfig struct {
+	Level   SentinelType
+	MaxSize int64
+	Compact func([]*parts.Part) (*parts.Part, int64, int64, error)
+}
+
+type LSMOption func(*LSM)
+
+func LSMWithRegistry(reg prometheus.Registerer) LSMOption {
+	return func(l *LSM) {
+		l.metrics.reg = reg
+	}
+}
+
+func LSMWithLogger(logger log.Logger) LSMOption {
+	return func(l *LSM) {
+		l.logger = logger
+	}
+}
+
+// NewLSM returns an LSM-like index of len(levels) levels.
+func NewLSM(prefix string, levels []*LevelConfig, options ...LSMOption) (*LSM, error) {
+	if err := validateLevels(levels); err != nil {
+		return nil, err
+	}
+
+	lsm := &LSM{
+		prefix:     prefix,
+		levels:     NewList(L0),
+		sizes:      make([]*atomic.Int64, len(levels)),
+		configs:    levels,
+		compacting: &atomic.Bool{},
+		logger:     log.NewNopLogger(),
+		metrics: &LSMMetrics{
+			reg: prometheus.NewRegistry(),
+		},
+	}
+
+	for _, opt := range options {
+		opt(lsm)
+	}
+
+	for i := range lsm.sizes {
+		lsm.sizes[i] = &atomic.Int64{}
```

Comment:
nit: if you make the `sizes` slice a slice of values, the atomic is already correctly initialized.